text/0001-reactive-assignments.md

```diff
@@ -0,0 +1,1068 @@
+
+To summarise, there are **3 simple rules** for understanding the code in a Svelte component's `<script>` block:
+
+* The code runs once for each instantiated component
+* Variable assignments trigger view updates
+* `export`ed values form the component's contract with the outside world
+
+
+### Lifecycle
+
+Many components need to respond to *lifecycle events*. These are currently expressed in Svelte via four [lifecycle hooks](https://svelte.technology/guide#lifecycle-hooks) — `onstate`, `oncreate`, `onupdate` and `ondestroy`.
+
+The `oncreate` hook runs *after* the initial render, meaning there is no way (in Svelte 2) to run setup code immediately upon instantiation — something this proposal solves. There is still a use for it, as we'll see, but in many cases `oncreate` will be unnecessary.
+
+That setup code could include work that needs to be undone when the component is removed. For that, we import the `onDestroy` lifecycle function (no longer called 'hooks', to avoid confusion with React Hooks which have a fundamentally different mechanism):
+
+```html
+<script>
+  import { onDestroy } from 'svelte';
+  import { format_time } from './helpers.js';
+
+  let time = new Date();
+
+  const interval = setInterval(() => {
+    time = new Date();
+  }, 1000);
+
+  onDestroy(() => {
+    clearInterval(interval);
+  });
+</script>
+
+<p>The time is {format_time(time)}</p>
+```
+
+The `onDestroy` callback is associated with the component instance because it is called during instantiation; no compiler magic is involved. Beyond that (if it is called outside instantiation, an error is thrown), there are no rules or restrictions as to when a lifecycle function is called. Reusability is therefore straightforward:
+
+```js
+// helpers.js
+import { onDestroy } from 'svelte';
+
+export function useInterval(fn, ms) {
+  const interval = setInterval(fn, ms);
+  onDestroy(() => clearInterval(interval));
+  fn();
+}
+```
+
+```html
+<script>
+  import { useInterval, formatTime } from './helpers.js';
+
+  let time;
+  useInterval(() => time = new Date(), 1000);
+</script>
+
+<p>The time is {formatTime(time)}</p>
+```
+
+> This might seem less ergonomic than React Hooks, whereby you can do `const time = useCustomHook()`. The payoff is that you don't need to run that code on every single state change, and it's easier to see which values in a component are subject to change, and *when* a specific value is changing.
+
+There are three other lifecycle functions required — `onMount` (similar to `oncreate` in Svelte v2), `beforeUpdate` (similar to `onstate`) and `afterUpdate` (similar to `onupdate`):
+
+```html
+<script>
+  import { beforeUpdate, afterUpdate, onMount } from 'svelte';
+
+  export let foo;
+
+  beforeUpdate(() => {
+    // this callback runs whenever props change
```

Comment:
I believe this also runs when reactive assignments are made, so I think "props" here may be confusing since I understand props to only be those values that are exported.
```suggestion
    // this callback runs whenever props or reactive variables change
```